### DISCUSSION TITLE
Help: NEXTAUTH_URL environment variable missing on Vercel suddenly

### DISCUSSION THREAD

**Discussion by adiliqbl** [nextauthjs/next-auth#1415] [Help] [answered]
NEXTAUTH_URL env variable suddenly missing on Vercel after a deployment (_with no changes to NEXTAUTH_URL_). Here is the link to the website [https://swanvo.com](https://swanvo.com)

The website has been up and running on vercel since 3 weeks now and a small deployment resulted in 500 error. I have tried deploying previous commits, resetting master branch and testing with vercel cli (which is working as expected). Even setting environment variables in vercel dashboard didn't work.

Here's the function log from vercel:
```
[GET] /api/auth/providers
13:21:48:05
2021-03-02T08:21:49.251Z eedb719c-b3f4-4474-9792-053e7643ef93 WARN [next-auth][warn][nextauth_url]
https://next-auth.js.org/warnings#nextauth_url NEXTAUTH_URL environment variable not set
2021-03-02T08:21:49.348Z eedb719c-b3f4-4474-9792-053e7643ef93 ERROR TypeError: Cannot read property 'replace' of undefined
at Object.qvzy (/var/task/.next/server/pages/api/auth/[...nextauth].js:1704:52)
at __webpack_require__ (/var/task/.next/server/pages/api/auth/[...nextauth].js:23:31)
at Module.Y/DP (/var/task/.next/server/pages/api/auth/[...nextauth].js:638:21)
at __webpack_require__ (/var/task/.next/server/pages/api/auth/[...nextauth].js:23:31)
at Object.5 (/var/task/.next/server/pages/api/auth/[...nextauth].js:201:18)
at __webpack_require__ (/var/task/.next/server/pages/api/auth/[...nextauth].js:23:31)
at /var/task/.next/server/pages/api/auth/[...nextauth].js:91:18
at Object.<anonymous> (/var/task/.next/server/pages/api/auth/[...nextauth].js:94:10)
at Module._compile (internal/modules/cjs/loader.js:999:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
RequestId: eedb719c-b3f4-4474-9792-053e7643ef93 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

I believe it has something to do with vercel caching or am I missing something? Any help would be greatly appreciated, thanks :)

**Comment by balazsorban44** [comment]
So I post my answer from https://github.com/vercel/next.js/issues/22667 as well, and let's continue this discussion here instead :wink: 

Would be interesting to see some kind of a reproduction (or at least your .env and [...nextauth].js file) though and some verification that you correctly set the NEXTAUTH_URL. The above error occurs when the value is undefined:

nextauthjs/next-auth@d86609a/src/server/index.js#L20-L22

You shouln't be needing to set env in next.config.js either.

**Reply by adiliqbl** [reply]
Thanks for reply @balazsorban44. Here's the config:

.env.production
```
NEXTAUTH_URL = "https://swanvo.com"
```

Added these to see if it fixes and I printed out the `procee.env` in next.config.js and found that all variables were present on vercel build logs.

next.config.js
```
module.exports = {
	env: {
		NEXT_PUBLIC_WEBSITE_URL: process.env.NEXT_PUBLIC_VERCEL_URL || process.env.NEXTAUTH_URL,
		NEXTAUTH_URL: process.env.NEXT_PUBLIC_VERCEL_URL || process.env.NEXTAUTH_URL
	}
}
```

[...nextauth].js
```javascript
export default NextAuth({
	providers: [
		Providers.Google({
			clientId: process.env.GOOGLE_CLIENT_ID,
			clientSecret: process.env.GOOGLE_CLIENT_SECRET,
			scope:
				'https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/calendar.readonly',
			authorizationUrl:
				'https://accounts.google.com/o/oauth2/v2/auth?prompt=consent&access_type=offline&response_type=code'
		})
	],
	adapter: FirebaseAuth,
	callbacks: {
		session: async (session, user) => {
			if (user && session.user) {
				session.user.id = user.userId;
				session.user.username = user.username;
				session.accessToken = user.accessToken;
				session.refreshToken = user.refreshToken;
			}
			return session;
		},
		jwt: async (token, user, account, profile, isNewUser) => {
			if (user) {
				token.userId = user.id;
				token.username = user.username;
			}
			if (account) {
				token.accessToken = account.accessToken;
				token.refreshToken = account.refreshToken;
				token.accessTokenExpires = account.accessTokenExpires;
			}
			if (user && account) {
				return token;
			} else if (Date.now() < token.accessTokenExpires) {
				return token;
			} else {
				// access token has expired, try to update it
				return refreshAccessToken(token);
			}
		},
		/**
		 * @param  {string} url      URL provided as callback URL by the client
		 * @param  {string} baseUrl  Default base URL of site (can be used as fallback)
		 * @return {string}          URL the client will be redirect to
		 */
		redirect: async (url, baseUrl) => {
			return url.startsWith(baseUrl) ? Promise.resolve(url) : Promise.resolve(baseUrl);
		}
	},
	secret: process.env.COOKIE_SECRET,
	session: {
		jwt: true,
		maxAge: 30 * 24 * 60 * 60, // 30 days
		updateAge: 24 * 60 * 60 // 24 hours
	},
	jwt: {
		secureCookie: process.env.NODE_ENV === 'production',
		encrypt: true
	},
	pages: {
		signIn: Links.LOGIN,
		signOut: Links.Private.LOGOUT,
		error: '/auth/error',
		verifyRequest: Links.VERIFY_USER, 
		newUser: null
	}
});
```

**Reply by Manubi** [reply]
you don't need to set the .env's in your next.config. Delete them there. What I find is the best workflow: go to your vercel dashboard and go to the project settings where you can add every environment variable for every different environment (dev, preview, prod). Then put this in your package json "vercel env pull && yarn dev" to start the def environment. It will pull the latest env variables and start the dev environment. Especially nice if you have more people working on the project who can edit env's.

btw: the NEXT_PUBLIC_ prefix makes it available in the browser.
without NEXT_PUBLIC the env will only be available server-side.

**Reply by adiliqbl** [reply]
I updated the config as you asked and still can't make it to work on vercel. For the meanwhile, we've deployed the website to Heroku and will play around with vercel to get it working. Will update here if I'm able to find the underlying issue.

**Reply by Manubi** [reply]
Does any off the envs work locally? You also can use the vercel cli to add the env’s.

**Reply by alexcarcoana** [reply]
> you don't need to set the .env's in your next.config. Delete them there. What I find is the best workflow: go to your vercel dashboard and go to the project settings where you can add every environment variable for every different environment (dev, preview, prod). Then put this in your package json "vercel env pull && yarn dev" to start the def environment. It will pull the latest env variables and start the dev environment. Especially nice if you have more people working on the project who can edit env's.
> 
> btw: the NEXT_PUBLIC_ prefix makes it available in the browser.
> without NEXT_PUBLIC the env will only be available server-side.

Out of curiosity, what variables need to be available to the browser and which should be kept serverside?

**Reply by balazsorban44** [reply]
https://nextjs.org/docs/basic-features/environment-variables

secrets for example (like database connections, provider client secrets, etc) should never be exposed to a browser

**Reply by khill-fbmc** [reply]
> ... put this in your package json "vercel env pull" ... 

I stumbled here from somewhere, and found this command super helpful. For anyone else using TypeScript, running this little script right after will generate a `process.env.d.ts` with keys for all the envvars.

```node
import { readFile, writeFile } from "node:fs/promises";

import * as dotenv from "dotenv";

const inputEnvFile = ".env";
const outputFile = "process.env.d.ts";

(async function () {
  const start = new Date().getTime();
  const content = await readFile(`${process.cwd()}/${inputEnvFile}`);
  const config = dotenv.parse(content);
  const keys = Object.keys(config).map(k => `      ${k}: string;`);
  const output = `// This file was generated
declare global {
  namespace NodeJS {
    interface ProcessEnv {
${keys.join("\n")}
    }
  }
}
export {};
`;

  await writeFile(`${process.cwd()}/${outputFile}`, output);
  const finish = new Date().getTime();
  const time = `[${String(finish - start).concat("ms")}]`;
  console.log(`✅  Generated \x1b[1m${outputFile}\x1b[0m \x1b[2m${time}\x1b[0m`);
})();
```

The generated file...

![image](https://user-images.githubusercontent.com/107003416/220971283-86062de0-d8fa-4301-89e1-9c7e55edc9a1.png)

Hooray! Types!

![image](https://user-images.githubusercontent.com/107003416/220971568-7f473720-7dc5-41be-b9a0-b75115e2fba0.png)

**Comment by adiliqbl** [comment, marked as answer]
The issue was related to Vercel's Next.js Builder. Support asked me to add FORCE_RUNTIME_TAG=canary environment variable, as they've fixed internal bug, which fixed the issue.

Closing this ticket as it's unrelated to next-auth. Highly appreciate quick help, been a rough day :)

**Reply by balazsorban44** [reply]
happy that it worked out well!
